packages/amplify-cli-core/src/permissionBoundaryState.ts

```diff
@@ -0,0 +1,55 @@
+import { stateManager } from './state-manager';
+import _ from 'lodash';
+import { $TSObject } from '.';
+
+export const getPermissionBoundaryArn = (env?: string): string | undefined => {
+  try {
+    const preInitTpi = (global as any).preInitTeamProviderInfo;
+    const teamProviderInfo = preInitTpi ?? stateManager.getTeamProviderInfo();
+    // if the pre init team-provider-info only has one env (which should always be the case), default to that one
+    if (preInitTpi && Object.keys(preInitTpi).length === 1 && !env) {
+      env = Object.keys(preInitTpi)[0];
+    }
+    return _.get(teamProviderInfo, teamProviderInfoObjectPath(env)) as string | undefined;
+  } catch (err) {
+    // uninitialized project
+    return undefined;
+  }
+};
+
+/**
+ * Stores the permission boundary ARN in team-provider-info
+ * If teamProviderInfo is not specified, the file is read, updated and written back to disk
+ * If teamProviderInfo is specified, then this function assumes that the env is not initialized
+ *    In this case, the teamProviderInfo object is updated but not written to disk. Instead "preInitTeamProviderInfo" is set
+ *    so that subsequent calls to getPermissionBoundaryArn will return the permission boundary arn of the pre-initialized env
+ * @param arn The permission boundary arn. If undefined or empty, the permission boundary is removed
+ * @param env The Amplify env to update. If not specified, defaults to the current checked out environment
+ * @param teamProviderInfo The team-provider-info object to update
+ */
+export const setPermissionBoundaryArn = (arn?: string, env?: string, teamProviderInfo?: $TSObject): void => {
+  let tpiGetter = () => stateManager.getTeamProviderInfo();
+  let tpiSetter = (tpi: $TSObject) => {
+    stateManager.setTeamProviderInfo(undefined, tpi);
+    delete (global as any).preInitTeamProviderInfo; // avoids a potential edge case where set permissions is called again w/o tpi
```

Comment:
Why are we using a global variable? I see this being used only here and no where else. Can't a variable in closure suffice?

Reply:
>  The variable could be scoped to the module but there's still a small chance that two packages that depend on this one would resolve to a different version and lead to some difficult to debug issues.

In that case do they use the setter or do they directly set `preInitTeamProviderInfo` global variable? Could you point me to the code where its being set other than this module?

Reply:
it's not retrieved or set anywhere except in this module. But if one package imports this module to set the value and another package imports the module to retrieve the value it's possible that the import won't resolve to the same instance of the module leading to unexpected behavior

Reply:
We have other places that would potentially have this issue (feature flags for example) so I'm okay with making it a module-scoped variable if you think that's the better choice

Reply:
> it's not retrieved or set anywhere except in this module. But if one package imports this module to set the value and another package imports the module to retrieve the value it's possible that the import won't resolve to the same instance of the module leading to unexpected behavior

This happens only if the required cache is cleared, which should not be done. In normal cases, once the package is imported in one module, other module importing it wont reset the values. So we should be fine.